### PR TITLE
Integration with the testbed-develop Branch

### PR DESCRIPTION
**Pull request by namitshah**
Please review the files for merging.

### CODE REVIEW

File: WAFL-Testbed/ctrl/SubsetGenerator.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        return len(self.data)
+
+    def __getitem__(self, idx):
+        return self.data[idx]
+
+
+# Number of WAFL Agents.
+N = 10
+# Path to the Filter File.
+FILTER_PATH = "../data/noniid_filter/filter_r90_s01.pt"
```
Comment (babcs2035) [inline WAFL-Testbed/ctrl/SubsetGenerator.py]: The filter file should exist?

File: WAFL-Testbed/ctrl/SubsetGenerator.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    root=DATA_PATH,
+    train=True,
+    download=True,
+    transform=torchvision.transforms.transforms.ToTensor(),
+)
+indices = torch.load(FILTER_PATH)
+
+for i in range(N):
+    # Materializing the Dataset.
+    subset = materialize_dataset(torch.utils.data.dataset.Subset(trainset, indices[i]))
```
Comment (babcs2035) [inline WAFL-Testbed/ctrl/SubsetGenerator.py]: You should replace torch.utils.data.dataset.Subset with torch.utils.data.Subset, as Subset is directly under torch.utils.data.

File: WAFL-Testbed/wafl/src/common/main.py [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+        model_instance_path: str,
+        contact_pattern_path: str,
+        model_sharing: ModelSharingUtils,
+        ctrl_tcp: CTRL_TCP,
+    ) -> None:
+        """
+        Initialize the learning process instance.
+        """
+        torch.random.manual_seed(1)
+        dataset = pickle.load(open(dataset_path, "rb"))
```
Comment (babcs2035) [inline WAFL-Testbed/wafl/src/common/main.py]:
```suggestion
        with open(dataset_path, "rb") as f:
             dataset = pickle.load(f)
```

File: WAFL-Testbed/wafl/src/common/main.py [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+        self.criterion = nn.CrossEntropyLoss()
+        self.optimizer = optim.Adam(self.net.parameters(), lr=ModelLearningUtils.cLEARNING_RATE)
+        self.logger = logging.getLogger("ModelLearningUtils")
+        self.logger.info("Initialized the Model Learning Utils instance.")
+
+    def get_neighbour_list(self, five_digit_number_str: str = "99999") -> List[int]:
+        """
+        Return the neighbour list for the current epoch from the contact pattern file.
+        """
+        neighbour_list = self.neighbour_map[int(five_digit_number_str)][str(self.model_sharing.name)]
```
Comment (babcs2035) [inline WAFL-Testbed/wafl/src/common/main.py]:
```suggestion
        try:
            neighbour_list = self.neighbour_map.get(int(five_digit_number_str), {}).get(str(self.model_sharing.name), [])
        except Exception as exc:
            self.logger.error(f"Error retrieving neighbour list: {str(exc)[:100]}...")
```

File: WAFL-Testbed/wafl/src/common/main.py [hunk 5]
```diff
@@ -0,0 +1,677 @@
...
+        """
+        self.logger.info(f"🎯 Beginning the WAFL-Learning Epoch: {five_digit_number_str}")
+        SUCCESS = False
+        try:
+            neighbours = self.get_neighbour_list(five_digit_number_str)
+            n_nbr = len(neighbours)
+            local_model = copy.deepcopy(self.net.state_dict())
+            for neighbour in neighbours:
+                received_model = self.model_sharing.request_model_from_peer(
+                    self.ctrl_tcp.get_device_ip(neighbour), "&purpose=testing"
```
Comment (babcs2035) [inline WAFL-Testbed/wafl/src/common/main.py:150]:
```suggestion
                neighbour_ip = self.ctrl_tcp.get_device_ip(neighbour)
                if neighbour_ip is None:
                    self.logger.warning(f"Skipping neighbour {neighbour}: IP address not found.")
                    continue
                received_model = self.model_sharing.request_model_from_peer(neighbour_ip, "&purpose=testing")
```

File: WAFL-Testbed/wafl/src/common/main.py [hunk 6]
```diff
@@ -0,0 +1,677 @@
...
+            )
+            return None
+
+    def setup_local_wafl_node(self, device_name: int) -> None:
+        """
+        Sets up the node with the given device name.
+        This function should be called before starting the WAFL model training.
+        """
+        self.logger.info(f"Setting up the node with device name: {device_name}")
+        self.model_sharing = ModelSharingUtils(device_name, self.get_device_ip(device_name), self.p2p_port, 10.0)
```
Comment (babcs2035) [inline WAFL-Testbed/wafl/src/common/main.py:504]:
```suggestion
        device_ip = self.get_device_ip(device_name)
        if device_ip is None:
            self.logger.error(f"Cannot set up node: IP address for device {device_name} not found.")
            return
        self.model_sharing = ModelSharingUtils(device_name, device_ip, self.p2p_port, 10.0)
```

File: WAFL-Testbed/wafl/src/common/main.py [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+        # --- Update status at the end of the epoch ---
+        self.status_logs.append(f"Epoch completion successful: {FLAG}")
+        self.is_epoch_running = False
+        # ---
+        return FLAG
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.DEBUG)
+    comm_interface = CTRL_TCP("../config/config_local.json")
```
Comment (babcs2035) [inline WAFL-Testbed/wafl/src/common/main.py]: The config file name should be "config.json", not "config_local.json"?

**Review by babcs2035 — CHANGES_REQUESTED** [review]
I commented mainly about error handling.

File: WAFL-Testbed/wafl/src/common/main.py (hunk 4 above)
Comment (namitshah) [reply WAFL-Testbed/wafl/src/common/main.py]: Hi, I did not add error handling for this since an exception would be raised only if the contact pattern file was inconsistent, in which case it would be detected and handled by the except block of the calling function.

The contact pattern files are lists of dicts (of neighbor lists), so I'm modifying this snippet.
Also, in my opinion it is better to raise an exception (no value for the key in the dict) rather than return an empty neighbor list since that would mask any issue with the contact pattern file. But that would depend on the structure of the contact pattern file, so better to just return an empty list for now.

File: WAFL-Testbed/wafl/src/common/main.py (hunk 5 above)
Comment (namitshah) [reply WAFL-Testbed/wafl/src/common/main.py:150]: In my opinion, is it not better to let the WAFL agent be stuck if it is unable to retrieve any of its neighbors' IP addresses? Since this is a testbed for research projects, and we are aiming for consistent and reproducible behavior, letting execution continue with a warning even though there's a major configuration issue with the device IPs does not seem ideal, right?

Comment (babcs2035) [reply WAFL-Testbed/wafl/src/common/main.py:150]: I agree with your idea, the exception should be raised on such cases.

File: WAFL-Testbed/ctrl/SubsetGenerator.py (hunk 1 above)
Comment (namitshah) [reply WAFL-Testbed/ctrl/SubsetGenerator.py]: Hi, the path strings in the SubsetGenerator.py file are placeholders, the actual values would depend on the specifications of the project (the non-IID filter file chosen, the location of the raw MNIST database,  etc). Since this file will be executed before deploy.sh, and are not to be sent to the WAFL agents, these can be set at runtime.

File: WAFL-Testbed/wafl/src/common/main.py (hunk 7 above)
Comment (namitshah) [reply WAFL-Testbed/wafl/src/common/main.py]: There was another config_local.json file. Am removing it from config and am moving the config.json file from src/common to config/common

File: WAFL-Testbed/wafl/src/common/main.py (hunk 6 above)
Comment (namitshah) [reply WAFL-Testbed/wafl/src/common/main.py:504]: Similar to the suggested change at line #140.
This is a critical error and would currently be handled similarly by the socket_listener_thread method.
Let me know if this should be added though, I'll add it, though it's not necessary imo.

File: WAFL-Testbed/wafl/src/common/main.py (hunk 7 above)
Comment (babcs2035) [reply WAFL-Testbed/wafl/src/common/main.py]: What is the difference between them?

**Comment by namitshah** [comment]
@babcs2035,
Thank you for the comprehensive review.
I have committed some of the suggested changes and have added comments where I think some more discussion might be needed. Let me know if there are any other suggestions.

File: WAFL-Testbed/wafl/src/common/main.py (hunk 5 above)
Comment (namitshah) [reply WAFL-Testbed/wafl/src/common/main.py:150]: It will be handled by the request_model_from_peer method, which won't let the execution move forward if something like this happens and will keep raising exceptions.

File: WAFL-Testbed/wafl/src/common/main.py (hunk 7 above)
Comment (namitshah) [reply WAFL-Testbed/wafl/src/common/main.py]: They were identical, hence have removed the local one. We used it initially for testing and it just remained in the directory even after the other config file was added. Also the wafl/config/common/config.json file is actually for Device 0, so just a placeholder. But this doesn't matter since the actual config files are anyway being generated by ctrl/main.py.

File: WAFL-Testbed/wafl/src/common/main.py [hunk 8]
```diff
@@ -0,0 +1,682 @@
...
+            if sub_command == "SELF":
+                return self._self_learn(five_digit_number_str)
+            elif sub_command == "WAFL":
+                return self._wafl_learn(five_digit_number_str)
+            else:
+                self.logger.warning(f"Unknown BEGIN subcommand: {sub_command} in command: {command_str}")
+                return False
+        else:
+            self.logger.warning(f"Unknown command received: {command_str}")
+            return False
```
Comment (nat0391) [inline WAFL-Testbed/wafl/src/common/main.py]: Execution server should respond with "OK" or "ERROR" to the "BEGIN" command.

File: WAFL-Testbed/wafl/src/common/main.py [hunk 9]
```diff
@@ -0,0 +1,682 @@
...
+        elif main_command == "BEGIN":
+            # This is a sample command structure; you might need to adjust it.
+            # Example: BEGIN-SELF-00001 or BEGIN-WAFL-00002-[1,2,3]
+            sub_command = parts[1]
+            five_digit_number_str = parts[2]
+
+            if sub_command == "SELF":
+                return self._self_learn(five_digit_number_str)
+            elif sub_command == "WAFL":
+                return self._wafl_learn(five_digit_number_str)
```
Comment (nat0391) [inline WAFL-Testbed/wafl/src/common/main.py]: With this implementation, is it  impossible to receive command from control server while execute learning?
I suggest threading the learning part.

File: WAFL-Testbed/wafl/src/common/main.py [hunk 10]
```diff
@@ -0,0 +1,682 @@
...
+
+        if main_command == "KILL":
+            if len(parts) != 1:
+                self.logger.warning(f"Incorrect form of KILL command: {command_str}")
+                return False
+            # Handle the KILL command by setting flags to terminate loops.
+            self.logger.info("KILL command received. Shutting down listeners.")
+            self.fLISTENER_ACTIVE = False
+            self.model_sharing.fLISTENER_ACTIVE = False
+            return True
```
Comment (nat0391) [inline WAFL-Testbed/wafl/src/common/main.py]: Execution server should respond with "OK" or "ERROR" to the "KILL" command.
Also, I think it's necessary to check the threads(e.g. P2P listener) are actually terminated before response.

**Review by nat0391 — APPROVED** [review]
I reviewed and left some comments.

**Comment by Ta2305** [comment]
I fixed problems in wafl/main.py pointed out in the review.
